### PR TITLE
Submit CCSMB-13: Stereo DFPWM Audio Protocol

### PR DESCRIPTION
**Pull request by drucifer-sc**
Submitting CCSMB-13: Stereo DFWPM Audio Protocol

### CODE REVIEW

File: Standards/CCSMB-13.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+Stereo DFPWM Audio Protocol (SDAP) has two functional rulesets depending on the modem used:
+| Modem Type | Transmission Mode |
+| ---------- | ----------------- |
+| Ender      | Full-Power        |
+| Wireless   | Low-Power         |
+| Wired      | Low-Power/Wired   |
+
+
+*Important: SDAP does not use Rednet! SDAP is a direct-modem protocol! Broadcasting SDAP packets via Rednet on a public server may result in bad things happening and I won't be held responsbile. Always make sure you're using direct modem calls, not Rednet!*
```
Comment (EmmaKnijn) [inline Standards/CCSMB-13.md]: I propose we define network encapsulation and such so we can either exclude this from it or make sure it wont conflict.

File: Standards/CCSMB-13.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+Example modem transmit snippet to give you an idea of where things go: `modem.transmit(65500,1337,string.pack("s1s1s2","KDRU Drucifer's Private Reserves","Artist - Title",leftChannel:sub(1,6000)..rightChannel:sub(1,6000))` where `leftChannel` and `rightChannel` contain DFPWM encoded aduio buffers of 6000 or more bytes.
```
Comment (EmmaKnijn) [inline Standards/CCSMB-13.md]: This could probably better be a full codeblock instead of an inline codeblock

**Review by EmmaKnijn — CHANGES_REQUESTED** [review]
I think we should include the regular RFC 2119.
We might also want to think about if we integrate this into CCSMB-5 or keep it as a seperate standard.

File: Standards/CCSMB-13.md (hunk 1 above)
Comment (drucifer-sc) [reply Standards/CCSMB-13.md]: I defined modem usage and split off this notice under the subheader Transmission medium(s)

File: Standards/CCSMB-13.md (hunk 2 above)
Comment (drucifer-sc) [reply Standards/CCSMB-13.md]: Added 2, count 'em, two codeblocks with expounded code samples.

**Review by EmmaKnijn — APPROVED** [review]

**Comment by MCJack123** [comment]
There's a lot of overlap here with the existing (and yet to be approved) CCSMB-5. And I don't see much benefit that this brings over that proposal.

- CCSMB-5 has basically the same capabilities as this proposal - it has a station field, a metadata field, and support for multiple channels of audio, all over normal modems. However, it also supports 1-8 channels with variable-length chunks instead of just two fixed chunks, and is already decoded (more about that below).
- I don't understand the benefit of using packed string data over tables. This may very slightly reduce server impact, but in exchange, it makes the format difficult to extend without breaking existing interpreters. It also introduces artificial limits that wouldn't need to exist if it was just a table.
- CCSMB-5 has much richer metadata, which is suitable for programs to display as they wish; whereas this proposal only provisions for a station name + a single title field. CCSMB-5 being a table also makes this easier to implement - stations can skip fields that they don't need to send (IIRC).
- Broadcasting DFPWM instead of decoded PCM will put extra processing impact on the MC server. Instead of decoding the data a single time on the server computer, each client must decode the data individually, which will really add up with lots of clients!

**Comment by drucifer-sc** [comment]
There's a lot of overlap here with the existing (and yet to be approved) CCSMB-5. And I don't see much benefit that this brings over that proposal. **--Cool, got it, will close and destroy this.**

CCSMB-5 has basically the same capabilities as this proposal - it has a station field, a metadata field, and support for multiple channels of audio, all over normal modems. However, it also supports 1-8 channels with variable-length chunks instead of just two fixed chunks, and is already decoded (more about that below). **That's great but a table does not transmit well over other transport mediums. Mine was meant to be transit agnostic and survive being pushed over HTTP, WS, or modem.**
I don't understand the benefit of using packed string data over tables. This may very slightly reduce server impact, but in exchange, it makes the format difficult to extend without breaking existing interpreters. It also introduces artificial limits that wouldn't need to exist if it was just a table. **A packed string takes up LESS bytes than a table being passed in plain text. Simple.**
CCSMB-5 has much richer metadata, which is suitable for programs to display as they wish; whereas this proposal only provisions for a station name + a single title field. CCSMB-5 being a table also makes this easier to implement - stations can skip fields that they don't need to send (IIRC). **The station title field is intended to be broadcast at all times to ensure the station is identifiable, the title field is entirely flexible in the 255 character limit as to what you want to put there. Additionally, I was thinking of making it skippable (0 byte string).**
Broadcasting DFPWM instead of decoded PCM will put extra processing impact on the MC server. Instead of decoding the data a single time on the server computer, each client must decode the data individually, which will really add up with lots of clients! **Packed DFPWM was selected to reduce the transmit size of the modem packets and make things a little faster. Again, it was selected to reduce the overall packet size with the intention that this format would end up being transit agnostic.**

**Comment by dimaguy** [comment]
I don't see the need to destroy this given you've shown valid points and ccsmb-5 isnt even approved yet. Show some spine sir and let this battle continue until the heat death of the universe with the merit of your designed format

**Pull request closed without merging**